ocaml/xapi/xapi_clustering.ml

```diff
@@ -111,6 +133,45 @@ let assert_pif_attached_to ~__context ~host ~pIF =
           (pif_not_attached_to_host, [Ref.string_of pIF; Ref.string_of host])
       )
 
+let assert_pif_share_network ~__context ~cluster ~host ~pIF ~index =
+  match Db.Cluster.get_cluster_hosts ~__context ~self:cluster with
+  | [] ->
+      debug "%s: This is the first cluster_host being created, so valid PIF"
+        __FUNCTION__
+  | chs ->
+      if index = 0L then (
+        let all_PIFs =
+          List.map (fun ch -> Db.Cluster_host.get_PIF ~__context ~self:ch) chs
+        in
+        if common_network ~__context (pIF :: all_PIFs) |> Option.is_none then
+          raise
+            Api_errors.(
+              Server_error
+                ( pif_not_in_cluster_network
+                , [Ref.string_of pIF; Ref.string_of host; Int64.to_string 0L]
+                )
+            )
+      ) else
+        let all_PIFs =
+          List.map
+            (fun ch -> Db.Cluster_host.get_other_PIFs ~__context ~self:ch)
+            chs
+        in
+        let nth_PIFs =
+          List.map (fun pIFs -> List.assoc_opt index pIFs) all_PIFs
+        in
+        if List.exists Option.is_none nth_PIFs then
+          failwith "Missing PIF in one of the cluster hosts in the cluster" ;
+        let nth_PIFs = pIF :: List.map Option.get nth_PIFs in
+        if common_network ~__context (pIF :: nth_PIFs) |> Option.is_none then
```

Comment:
this function here can potentially cause lots of DB calls, although we support max 9 ips, including the main one, so the worst case would probably be on the order of 100 calls. I am open to suggestoins on how/whether/if necessary we could do better